File: LeetCode/[M] - 763. Partition Labels/solution.py
from typing import List


class Solution:
    def partitionLabels(self, s: str) -> List[int]:
        lastIdx = {char: i for i, char in enumerate(s)}
        res = []
        lastGroup = 0
        l, r = 0, 0
        for l, char in enumerate(s):
            r = max(r, lastIdx[char])
            if l == r:
                res.append(r - lastGroup + 1)
                lastGroup = l + 1
            l += 1
        return res

File: LeetCode/[M] - 763. Partition Labels/solution.go
package main

func partitionLabels(s string) []int {
	lstIdx := make(map[rune]int, len(s))
	for i, char := range s {
		lstIdx[char] = i
	}
	var res []int
	var lstGroup int
	var r int
	for l, char := range s {
		r = max(r, lstIdx[char])
		if l == r {
			res = append(res, r-lstGroup+1)
			lstGroup = l + 1
		}
		l++
	}
	return res
}
